### PR TITLE
feat: Move code from physical to logical plan for Projection

### PR DESCRIPTION
**Pull request by vpapavas**
### Description 

No changes made to the persistent query code path. Ideally, I wanted to re-use the existing code without special casing for pull queries but that was not possible since the schemas are handled so differently. Below I describe with examples their differences.

Assume pull query `SELECT * from agg WHERE STR=foo;` where table is created as 
```
CREATE TABLE agg AS SELECT STR, LATEST_BY_OFFSET(LONG) AS LONG FROM stream GROUP BY STR;"
```

#### Logical schema handling in pull query project code. 
There are 3 schemas we handling, the input schema, the intermediate schema used for codegen and the output schema. 

- The input is the materialization schema. For example, `mat.schema = STR STRING KEY, LONG BIGINT` 
- CodeGen is used only if the SELECT is not `*`. For CodeGen, an intermediate schema is created as follows: Check if projection contains system or key columns. If not, the intermediate schema is the input schema. If there are any of these columns, the input schema is extended with `inputSchema	
.withPseudoAndKeyColsInValue(windowed);` that basically copies the key and system columns (rowtime, windwostart and windowend) into the value of the schema.
- The output schema (the schema of the row produced by the project operator), takes as input the materialization schema and: if the projection is  `SELECT *`, add windowstart and windowend to key columns and keep value columns the same as input. If projection is not `SELECT *`, then process each select and if it is a key or windowstart and windowend add them to the key part else add them to the value part. In our example, output schema is `STR STRING KEY, LONG BIGINT`.

#### Logical schema handling in logical plan builder. 
Each node of the logical plan has a logical schema associated with it and has access to its child logical schema. For pull queries the leaf is always a single `DataSourceNode`

- The output schema of the `DataSourceNode` (for the same example query as above) is `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING`. This is the input to the Project operator, compare this to the `mat.schema`
- The output schema of the `FinalProjectNode` is `STR STRING KEY, STR STRING, LONG BIGINT` (it has the extra key column in the value)

#### Different queries that highlight how the schema handling of pull queries differs from persistent queries

1. `SELECT long from agg WHERE STR=foo`
-----> FinalProjectNode input schema: `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING`
-----> mat.schema = `STR STRING KEY, LONG BIGINT`
-----> intermediate schema = `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING`
-----> FinalProjectNode output schema=`STR STRING KEY, LONG BIGINT`
-----> pull query output schema= `LONG BIGINT`

2. `SELECT long, rowtime from agg WHERE STR=foo`
-----> FinalProjectNode input schema: `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING`
-----> mat.schema = `STR STRING KEY, LONG BIGINT`
-----> intermediate schema= `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING`
-----> Project output schema=`STR STRING KEY, LONG BIGINT, ROWTIME BIGINT`
-----> pull query output schema= `LONG BIGINT, ROWTIME BIGINT`

3. `SELECT long, windowstart, windowend from agg WHERE STR=foo` assuming table is windowed
-----> FinalProjectNode input schema: `STR STRING KEY, ROWTIME BIGINT, WINDOWSTART BIGINT, WINDOWEND BIGINT`
-----> mat.schema = `STR STRING KEY, LONG BIGINT, WINDOWSTART BIGINT, WINDOWEND BIGINT`
-----> intermediate schema= `STR STRING KEY, LONG BIGINT, ROWTIME BIGINT, STR STRING,  WINDOWSTART BIGINT, WINDOWEND BIGINT`
-----> Project output schema=`STR STRING KEY`
-----> pull query output schema= `LONG BIGINT,  WINDOWSTART BIGINT KEY, WINDOWEND BIGINT KEY`

#### Questions:

1. What is the correspondence between DataSource and Materialization? Why do their schemas differ? If the Materialization is a table, then shouldn't the schema of the DataSource match that of the table? I.e not have the rowtime and key in the value?

2. Why do the output schemas of the pull query and `FinalProjectNode` differ? Basically, it seems that the output schemas of persistent queries are different than that of pull queries? Moreover, the output schema of `FinalProjectNode` seems to have extra columns that are not in the projection. Why?

### Testing done 
Added new unit tests for building the pull query schema

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/FinalProjectNode.java
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/FinalProjectNode.java:115]: >  What is the correspondence between DataSource and Materialization? Why do their schemas differ? If the Materialization is a table, then shouldn't the schema of the DataSource match that of the table? I.e not have the rowtime and key in the value?
>
>    Why do the output schemas of the pull query and FinalProjectNode differ? Basically, it seems that the output schemas of persistent queries are different than that of pull queries? Moreover, the output schema of FinalProjectNode seems to have extra columns that are not in the projection. Why?

Off the top of my head (see bottom), I don't know the answers to these questions (and I'm not sure that without digging anyone on the team knows) but I think these are exactly the ones we need to answer (and document for the future! Ideally in the code...) before merging this pr. Then, we can see if we can change it to be a single code path for both persistent/pull queries.

Otherwise (if we can't merge the code paths), I think we need to have two different logical nodes since they (logically) represent different operations and construct different schemas (they can extend a common base so that it's clear which code they share and which is different). We shouldn't make the pull and persistent code more intertwined but with a nuanced difference as it will make it harder to develop them independently without context on the other.

FWIW, you can change the behavior of any logical node of the persistent queries so long as the QTT tests pass. Any old query will continue to work because the physical plans are persisted, not the logical ones, in the command topic.

---


> Moreover, the output schema of FinalProjectNode seems to have extra columns that are not in the projection.

My best guess on this one is that the rowkey-in-value/rowtime-in-value/windowstart/windowend columns are needed to handle the processing. At the moment, the way we've coded it up, most operations accept only the value and not the key (or other metadata information) - so we need to copy information from the key into the value. I've looked into changing this and I got pretty close, but the problem is that the Joiner code in streams doesn't yet have the signature that I need to fully get rid of it so I abandoned the change. I'll try to revisit it (and hopefully change that in streams if necessary) at some point.

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/ProjectOperator.java
Comment (AlanConfluent) [inline ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/ProjectOperator.java:92]: If you just apply the transform immediately, can you just call it directly rather than returning a function?

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/ProjectOperator.java
Comment (AlanConfluent) [inline ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/ProjectOperator.java:138]: Maybe this should be called getIntermediateRow to match the terminology of the logical node?

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java
Comment (AlanConfluent) [reply ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java]: I think this is outdated

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (AlanConfluent) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:141]: But that's fine because this intermediate schema is never exposed, right?  The output schema puts these fields correctly as keys.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (AlanConfluent) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:54]: This is just to use the SelectValueMapper, right?

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (AlanConfluent) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:239]: Does this mean ExpressionTypeManager expects it?  That's the main place where this schema is used is just to evaluate the type of the expression.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java:174]: Some of the checks about about for instance the group by are performed in the Analyzer and specifically, by the class `PullQueryValidator`. The rest of the checks about the WHERE clause will be performed when building the filter node and will be in the next PR

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:95]: Yeah, for the next PR that does the WHERE clause I would have to share this schema. I will build it once in the WHERE and then the PROJECT since it sits above, will be able to get if from it. If a WHERE doesn't exist, then the PROJECT will build it

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:141]: Yes, they are not exposed

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:54]: Correct

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:239]: Also, the `CodeGenRunner` uses the schema

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:42]: thank you for the javadoc!

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:47]: excuse me while I philosophize for a bit...

just wondering, is 'child' and 'parent' standard terminology in databases when discussing logical plans? I know we've standardized on it, but it always confuses me because the "child" is the "source node" which makes me feel like it's a "parent" 🤔 

What do you think about upstream/downstream nodes? That way it's clear that the direction of data is flowing from the source nodes (upstream) to the sink nodes (downstream)

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/transform/select/SelectValueMapperFactory.java
Comment (agavra) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/transform/select/SelectValueMapperFactory.java:38]: any reason this is no longer `final`?

**Review by agavra — COMMENTED** [review]
the structure lgtm! I didn't spend much time dissecting the functionality of the code though

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (AlanConfluent) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:236]: I guess I just meant that `schema` is only used with the `ExpressionTypeManager` and the comment is a bit misleading.

**Review by AlanConfluent — APPROVED** [review]

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java
Comment (vpapavas) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:47]: The plans (logical, physical) are trees hence the terminology root, child, parent. The children are the leaves and as we move up the tree we visit from child to parent. Information and rows is flowing from the leaves towards the root.

Honestly, sorry but I find the upstream/downstream more confusing :(

Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java:47]: 😂 haha i'll keep working at making it feel intuitive!

**Pull request merged**
